### PR TITLE
Add more inbounds and inlining

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR applies more `@inbounds` and inlining to
 - some of the spectral element operators (written for the cpu)
 - some of the dss kernels

### CODE REVIEW

**Comment by charleskawczynski** [comment]
bors try

File: src/Operators/spectralelement.jl [hunk 1]
```diff
@@ -1355,7 +1405,7 @@ function apply_operator(op::WeakCurl{(1, 2)}, space, slabidx, arg)
     else
         error("invalid return type")
     end
-    for j in 1:Nq, i in 1:Nq
+    @inbounds for j in 1:Nq, i in 1:Nq
```
Comment (simonbyrne) [inline src/Operators/spectralelement.jl:1408]: you don't need `@inbounds` here if you use `@propagate_inbounds`

Comment (charleskawczynski) [reply src/Operators/spectralelement.jl:1408]: Ah, interesting, I thought `@inbounds` was needed any time a new index was introduced. Good catch!

**Comment by charleskawczynski** [comment]
Interestingly, this doesn’t seem to have much of an impact. Any differences could easily be variation.

**Comment by charleskawczynski** [comment]
The CPU vector dss is the slowest of the spectral kernels, and using IntrospectionTools along with `julia --check-bounds=no`:

```julia
using Revise; using ClimaCore
include(joinpath(pkgdir(ClimaCore), "test", "Operators", "spectralelement", "benchmark_utils.jl"))
include(joinpath(pkgdir(ClimaCore), "test", "Operators", "spectralelement", "benchmark_kernels.jl"))
args = setup_kernel_args(["--device", "CPU"]);
kernel_vector_dss!(args)
using IntrospectionTools
@code_summary kernel_vector_dss!(args)
```
shows that we're currently  generating the same native code as if we used `julia --check-bounds=no`, so these changes are not needed.

**Comment by charleskawczynski** [comment]
Actually, the allocation job shows that this fixes some allocation sites. So perhaps we should reconsider.

**Comment by charleskawczynski** [comment]
Superseded by #1338

**Pull request closed without merging**